numpy/_core/src/umath/string_buffer.h

```diff
@@ -1505,4 +1505,81 @@ string_expandtabs(Buffer<enc> buf, npy_int64 tabsize, Buffer<enc> out)
 }
 
 
+enum class JUSTPOSITION {
+    CENTER, LEFT, RIGHT
+};
+
+template <ENCODING enc>
+static inline npy_intp
+string_pad(Buffer<enc> buf, npy_int64 width, Buffer<enc> fill, JUSTPOSITION pos, Buffer<enc> out)
+{
+    size_t finalwidth = width > 0 ? width : 0;
+    size_t len = buf.num_codepoints();
+    if (len >= finalwidth) {
+        buf.buffer_memcpy(out, len);
+        out.buffer_fill_with_zeros_after_index(len);
+        return (npy_intp) len;
+    }
+
+    size_t left, right;
+    if (pos == JUSTPOSITION::CENTER) {
+        size_t pad = finalwidth - len;
+        left = pad / 2 + (pad & finalwidth & 1);
+        right = pad - left;
+    }
+    else if (pos == JUSTPOSITION::LEFT) {
+        left = 0;
+        right = finalwidth - len;
+    }
+    else {
+        left = finalwidth - len;
+        right = 0;
+    }
+
+    if (left == 0 && right == 0) {
+        buf.buffer_memcpy(out, len);
+        out.buffer_fill_with_zeros_after_index(len);
+        return len;
+    }
+
+    if (left > PY_SSIZE_T_MAX - len || right > PY_SSIZE_T_MAX - (left + len)) {
+        npy_gil_error(PyExc_OverflowError, "padded string is too long");
+        return -1;
+    }
+
+    if (left > 0) {
+        out += out.buffer_memset(*fill, left);
+    }
+    buf.buffer_memcpy(out, len);
+    out += len;
+    if (right > 0) {
+        out.buffer_memset(*fill, right);
+    }
+    return left + len + right;
+}
+
+
+template <ENCODING enc>
+static inline npy_intp
+string_zfill(Buffer<enc> buf, npy_int64 width, Buffer<enc> out)
+{
+    size_t finalwidth = width > 0 ? width : 0;
+
+    char fill = '0';
+    Buffer<enc> fillchar(&fill, 1);
+    npy_intp new_len = string_pad(buf, width, fillchar, JUSTPOSITION::RIGHT, out);
+
+    size_t offset = finalwidth - buf.num_codepoints();
+    Buffer<enc> tmp = buf + offset;
+
+    npy_ucs4 c = *tmp;
+    if (c == '+' || c == '-') {
```

Comment:
Should this include other unicode sign points? In particular, https://unicodeplus.com/U+2212?
Though I see that python doesn't do it, so probably better not:
```
In [9]: a = '\N{MINUS SIGN}1.0'

In [10]: a
Out[10]: '−1.0'

In [11]: a.zfill(10)
Out[11]: '000000−1.0'
```

Reply:
Yes, I checked that and CPython does not do it, so I decided against it.